### PR TITLE
refactor!: Modified all recipe interfaces to have exact return types.

### PR DESCRIPTION
**Pull request by girish946**
## Summary of change

- Modified EmailVerifyPostOkResponse and EmailVerifyPostInvalidTokenErrorResponse classes under recipe.emailvarification.interfaces.

- Modified EmailVerifyPostResponse class.
    
    This results in more readable code in recipe.emailverification.api.implementation. The code for email varification now looks like:
    
        response = await api_options.recipe_implementation.verify_email_using_token(token, user_context)
            if response.user:
                return EmailVerifyPostOkResponse(response.user)
            return EmailVerifyPostInvalidTokenErrorResponse()
    
- Added docstrings in EmailVerifyPostOkResponse, EmailVerifyPostInvalidTokenErrorResponse and email_verify_post method for email verification.

## Related issues

-   [Link to issue1 here](https://github.com/supertokens/supertokens-python/issues/75)

Signed-off-by: Girish Joshi <girish946@gmail.com>

### CODE REVIEW

File: supertokens_python/recipe/emailverification/api/implementation.py [hunk 1]
```diff
@@ -23,19 +23,39 @@
...
+
+                {
+                    "status": "EMAIL_VERIFICATION_INVALID_TOKEN_ERROR"
+                }
+        """
         response = await api_options.recipe_implementation.verify_email_using_token(token, user_context)
-        if response.is_ok:
-            if response.user is None:
-                raise Exception("Should never come here")
+        if response.user:
```
Comment (rishabhpoddar) [inline supertokens_python/recipe/emailverification/api/implementation.py:58]: This code should not change.

File: supertokens_python/recipe/emailverification/types.py [hunk 2]
```diff
@@ -12,8 +12,20 @@
...
+from typing import Any, Dict
+
 
 class User:
     def __init__(self, user_id: str, email: str):
         self.user_id = user_id
         self.email = email
+
+
+class STResponse(ABC):
```
Comment (rishabhpoddar) [inline supertokens_python/recipe/emailverification/types.py:25]: - Change name to `APIResponse`.
- This class should be in `supertokens_python/types.py`

File: supertokens_python/recipe/emailverification/interfaces.py [hunk 3]
```diff
@@ -133,31 +135,24 @@ def __init__(self, request: BaseRequest, response: BaseResponse, recipe_id: str,
...
-        self.user: Union[User, None] = user
-
-    def to_json(self) -> Dict[str, Any]:
-        return {
-            'status': self.status
-        }
+class EmailVerifyPostResponse(STResponse):
+    # TODO: This is a placeholder class for now, need to figure out
+    # correct hirarchy for Response class from supertokens
+    pass
```
Comment (rishabhpoddar) [inline supertokens_python/recipe/emailverification/interfaces.py:141]: This class is not actually needed?

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: CHANGELOG.md [hunk 4]
```diff
@@ -6,7 +6,12 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [unreleased]
 
-## [0.5.1] - 2022-03-02
+### Breaking Change
+- Refactored EmailVerifyPostOkResponse and EmailVerifyPostInvalidTokenErrorResponse classes under recipe.emailvarification.interfaces.
+- Removed EmailVerifyPostResponse class.
```
Comment (rishabhpoddar) [inline CHANGELOG.md:11]: Summarise this.

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: supertokens_python/recipe/emailverification/api/implementation.py (hunk 1 above)
Comment (girish946) [reply supertokens_python/recipe/emailverification/api/implementation.py]: Thanks for the review @rishabhpoddar .

I think we should change this.

At this point we always have the response which is an instance of `VerifyEmailUsingTokenResult`.
and we are ensuring that the user is a valid object of the class User before sending the response object which is
`EmailVerifyPostOkResponse`.
In the original code

```python
response = await api_options.recipe_implementation.verify_email_using_token(token, user_context)
    # we had to check is_ok since it notifies the http response from core which may contain an error in the status.
    if response.is_ok:
        # since we have not checked if the http response contains an error message
        # we need to validate if response.user is valid.  
        if response.user is None:
            
            raise Exception("Should never come here")
        return EmailVerifyPostOkResponse(response.user)
    return EmailVerifyPostInvalidTokenErrorResponse()
```
I've added `response.is_ok` validation now.
We can validate this in a more readable way 

```python
if response.is_ok:
    if response.user:
        return EmailVerifyPostOkResponse(response.user)
return EmailVerifyPostInvalidTokenErrorResponse()
```
If the `user` is `None`, we are anyways sending `EmailVerifyPostInvalidTokenErrorResponse`.

Please share your input.
Thanks.

File: supertokens_python/recipe/emailverification/types.py (hunk 2 above)
Comment (girish946) [reply supertokens_python/recipe/emailverification/types.py]: done.

File: CHANGELOG.md (hunk 4 above)
Comment (girish946) [reply CHANGELOG.md]: done.

File: supertokens_python/recipe/emailverification/api/implementation.py (hunk 1 above)
Comment (girish946) [reply supertokens_python/recipe/emailverification/api/implementation.py]: There was a little confusion initially.
Now it is resolved. updated this code.

File: CHANGELOG.md [hunk 5]
```diff
@@ -6,7 +6,11 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [unreleased]
 
-## [0.5.1] - 2022-03-02
+### Breaking Change
+- Refactored emailverify recipe interfaces.
```
Comment (rishabhpoddar) [inline CHANGELOG.md:10]: Put a bit more info about what the refactor is about. With an example. Also link to the relevant github issue

File: supertokens_python/recipe/emailverification/api/implementation.py [hunk 6]
```diff
@@ -13,65 +13,127 @@
...
     )
 
 from supertokens_python.recipe.emailverification.types import User
 from supertokens_python.recipe.session.asyncio import get_session
 
 
 class APIImplementation(APIInterface):
-    async def email_verify_post(self, token: str, api_options: APIOptions, user_context: Dict[str, Any]) -> EmailVerifyPostResponse:
+    async def email_verify_post(self, token: str, api_options: APIOptions, user_context: Dict[str, Any]) -> Union[EmailVerifyPostOkResponse, EmailVerifyPostInvalidTokenErrorResponse]:
+        """email_verify_post is an async mehtod that accepts
```
Comment (rishabhpoddar) [inline supertokens_python/recipe/emailverification/api/implementation.py:37]: Typo in "mehtod"

File: supertokens_python/recipe/emailverification/api/implementation.py [hunk 7]
```diff
@@ -13,65 +13,127 @@
...
+
+               - token: string,
+               - api_options: APIOptions,
+               - user_context: python dict
```
Comment (rishabhpoddar) [inline supertokens_python/recipe/emailverification/api/implementation.py:41]: This is not really needed as the function signature is very clear anyway. Instead, you might want to explain what this function is meant to do.

File: supertokens_python/recipe/emailverification/api/implementation.py [hunk 8]
```diff
@@ -13,65 +13,127 @@
...
+
+            if the user is successfully varified, the response could be as follows
```
Comment (rishabhpoddar) [inline supertokens_python/recipe/emailverification/api/implementation.py:43]: Typo in "varified"

File: supertokens_python/recipe/emailverification/api/implementation.py [hunk 9]
```diff
@@ -13,65 +13,127 @@
...
+
+                {
+                    "status": "OK",
+                    "user": {
+                      "id": "<usern-id>",
+                      "email": "<email-id>"
+                }
```
Comment (rishabhpoddar) [inline supertokens_python/recipe/emailverification/api/implementation.py:50]: The function doesn't return a JSON. So this is technically incorrect.

File: supertokens_python/recipe/emailverification/api/implementation.py [hunk 10]
```diff
@@ -13,65 +13,127 @@
...
+        # TODO: fix docstring, docstring should mention exact functioning of this method.
+        # TODO: this method should return an error response instead of an exception.
+
         session = await get_session(api_options.request)
-        if session is None:
-            raise Exception('Session is undefined. Should not come here.')
 
-        user_id = session.get_user_id(user_context)
-        email = await api_options.config.get_email_for_user_id(user_id, user_context)
+        if isinstance(session, SessionContainer):
```
Comment (rishabhpoddar) [inline supertokens_python/recipe/emailverification/api/implementation.py:87]: Checking `if session is None` is actually the correct thing to do here since the return type of `get_session` is `Union[SessionContainer, None]`. So undo this particular code change

File: supertokens_python/recipe/emailverification/api/implementation.py [hunk 11]
```diff
@@ -13,65 +13,127 @@
...
+            user_id = session.get_user_id(user_context)
+            email = await api_options.config.get_email_for_user_id(user_id, user_context)
+            is_verified = await api_options.recipe_implementation.is_email_verified(user_id, email, user_context)
+            return IsEmailVerifiedGetOkResponse(is_verified)
 
-        is_verified = await api_options.recipe_implementation.is_email_verified(user_id, email, user_context)
-        return IsEmailVerifiedGetOkResponse(is_verified)
+        raise Exception("Undefined Session")
 
     async def generate_email_verify_token_post(self, api_options: APIOptions, user_context: Dict[str, Any]) -> GenerateEmailVerifyTokenPostResponse:
```
Comment (rishabhpoddar) [inline supertokens_python/recipe/emailverification/api/implementation.py:95]: The return type of this function should also change to be a union of all the types that it actually returns. And then you don't need `GenerateEmailVerifyTokenPostResponse` anymore.

File: supertokens_python/recipe/emailverification/api/implementation.py [hunk 12]
```diff
@@ -13,65 +13,127 @@
...
+
+        In the case of failure:
+            {
+                "status": "EMAIL_ALREADY_VERIFIED_ERROR"
+            }
+        """
         session = await get_session(api_options.request)
-        if session is None:
-            raise Exception('Session is undefined. Should not come here.')
+        if isinstance(session, SessionContainer):
```
Comment (rishabhpoddar) [inline supertokens_python/recipe/emailverification/api/implementation.py:116]: No need to change how session checking works. So undo the code changes here

File: supertokens_python/recipe/emailverification/api/implementation.py [hunk 13]
```diff
@@ -13,65 +13,127 @@
...
+            user_id = session.get_user_id(user_context)
+            email = await api_options.config.get_email_for_user_id(user_id, user_context)
 
-        user_id = session.get_user_id(user_context)
-        email = await api_options.config.get_email_for_user_id(user_id, user_context)
+            token_result = await api_options.recipe_implementation.create_email_verification_token(user_id, email, user_context)
+            if token_result.is_email_already_verified:
```
Comment (rishabhpoddar) [inline supertokens_python/recipe/emailverification/api/implementation.py:121]: You should use isinstance to check the response type from that function call - as discussed earlier.

File: supertokens_python/recipe/emailverification/api/implementation.py [hunk 14]
```diff
@@ -13,65 +13,127 @@
...
 
-        user = User(user_id, email)
+            if isinstance(token_result, CreateEmailVerificationTokenOkResult):
 
-        if token_result.token is None:
-            raise Exception("Should never come here")
-        email_verify_link = (await api_options.config.get_email_verification_url(
-            user, user_context)) + '?token=' + token_result.token + '&rid' + api_options.recipe_id
+                # TODO: This validation is unnecessary since CreateEmailVerificationTokenOkResult needs a valid token.
+                if token_result.token:
```
Comment (rishabhpoddar) [inline supertokens_python/recipe/emailverification/api/implementation.py:129]: Pending TODO from your side (based on comments)

File: supertokens_python/recipe/emailverification/interfaces.py [hunk 15]
```diff
@@ -49,26 +50,18 @@ def __init__(self):
...
-        self.is_ok = False
-        self.is_email_verification_invalid_token_error = False
-        self.user = user
-
-
-class VerifyEmailUsingTokenOkResult(VerifyEmailUsingTokenResult):
+class VerifyEmailUsingTokenOkResult:
     def __init__(self, user: User):
-        super().__init__('OK', user)
+        self.status = 'OK'
```
Comment (rishabhpoddar) [inline supertokens_python/recipe/emailverification/interfaces.py:55]: The status value is not useful. Remove it

File: supertokens_python/recipe/emailverification/interfaces.py [hunk 16]
```diff
@@ -49,26 +50,18 @@ def __init__(self):
...
         self.is_ok = True
         self.is_email_verification_invalid_token_error = False
 
 
-class VerifyEmailUsingTokenInvalidTokenErrorResult(
-        VerifyEmailUsingTokenResult):
+class VerifyEmailUsingTokenInvalidTokenErrorResult:
     def __init__(self):
-        super().__init__('EMAIL_VERIFICATION_INVALID_TOKEN_ERROR', None)
+        self.status = 'EMAIL_VERIFICATION_INVALID_TOKEN_ERROR'
```
Comment (rishabhpoddar) [inline supertokens_python/recipe/emailverification/interfaces.py:63]: Status value is not useful. Remove it

File: supertokens_python/recipe/emailverification/interfaces.py [hunk 17]
```diff
@@ -49,26 +50,18 @@ def __init__(self):
...
-        self.user = user
-
-
-class VerifyEmailUsingTokenOkResult(VerifyEmailUsingTokenResult):
+class VerifyEmailUsingTokenOkResult:
     def __init__(self, user: User):
-        super().__init__('OK', user)
+        self.status = 'OK'
+        self.user = user
         self.is_ok = True
```
Comment (rishabhpoddar) [inline supertokens_python/recipe/emailverification/interfaces.py:57]: is_ok is not useful either.

File: supertokens_python/recipe/emailverification/interfaces.py [hunk 18]
```diff
@@ -49,26 +50,18 @@ def __init__(self):
...
         self.is_email_verification_invalid_token_error = False
```
Comment (rishabhpoddar) [inline supertokens_python/recipe/emailverification/interfaces.py:58]: is_email_verification_invalid_token_error is not required right - cause we will always use isinstance from now on

File: supertokens_python/recipe/emailverification/interfaces.py [hunk 19]
```diff
@@ -49,26 +50,18 @@ def __init__(self):
...
 
 
-class VerifyEmailUsingTokenInvalidTokenErrorResult(
-        VerifyEmailUsingTokenResult):
+class VerifyEmailUsingTokenInvalidTokenErrorResult:
     def __init__(self):
-        super().__init__('EMAIL_VERIFICATION_INVALID_TOKEN_ERROR', None)
+        self.status = 'EMAIL_VERIFICATION_INVALID_TOKEN_ERROR'
+        self.user = None
```
Comment (rishabhpoddar) [inline supertokens_python/recipe/emailverification/interfaces.py:64]: You don't need this type in this class. Not sure why this is assigned to None

File: supertokens_python/recipe/emailverification/interfaces.py [hunk 20]
```diff
@@ -49,26 +50,18 @@ def __init__(self):
...
         self.is_ok = False
         self.is_email_verification_invalid_token_error = True
```
Comment (rishabhpoddar) [inline supertokens_python/recipe/emailverification/interfaces.py:66]: Don't need is_* types anymore

File: supertokens_python/recipe/emailverification/interfaces.py [hunk 21]
```diff
@@ -49,26 +50,18 @@ def __init__(self):
         self.is_email_already_verified = True
```
Comment (rishabhpoddar) [inline supertokens_python/recipe/emailverification/interfaces.py:51]: Some types still need changing:
- RevokeEmailVerificationTokensResult
- UnverifyEmailResult
- GenerateEmailVerifyTokenPostResponse

File: supertokens_python/recipe/emailverification/interfaces.py [hunk 22]
```diff
@@ -133,59 +126,40 @@ def __init__(self, request: BaseRequest, response: BaseResponse, recipe_id: str,
...
-            'status': self.status
+            'status': 'EMAIL_VERIFICATION_INVALID_TOKEN_ERROR'
         }
 
 
-class IsEmailVerifiedGetOkResponse(IsEmailVerifiedGetResponse):
+class IsEmailVerifiedGetOkResponse(APIResponse):
     def __init__(self, is_verified: bool):
-        super().__init__('OK')
+        self.status = 'OK'
```
Comment (rishabhpoddar) [inline supertokens_python/recipe/emailverification/interfaces.py:162]: Be consistent with your changes. In other API return classes, you are not assigning a status field. So why do it here?

File: supertokens_python/recipe/emailverification/types.py [hunk 23]
```diff
@@ -12,7 +12,6 @@
 # License for the specific language governing permissions and limitations
 # under the License.
 
-
```
Comment (rishabhpoddar) [inline supertokens_python/recipe/emailverification/types.py:15]: Not sure why this change is here.

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: supertokens_python/recipe/emailverification/api/implementation.py (hunk 7 above)
Comment (girish946) [reply supertokens_python/recipe/emailverification/api/implementation.py:41]: This part is needed for the API documentation to be generated by `pdocs`. Currently this is a place holder. I need to dig into `pdocs` for improving this part.

File: supertokens_python/recipe/emailverification/interfaces.py (hunk 22 above)
Comment (girish946) [reply supertokens_python/recipe/emailverification/interfaces.py:158]: done.

File: supertokens_python/recipe/emailverification/types.py (hunk 23 above)
Comment (girish946) [reply supertokens_python/recipe/emailverification/types.py:15]: I think this is happening due to autopep8. It will be fixed when we use `black`.

File: supertokens_python/recipe/emailverification/interfaces.py (hunk 20 above)
Comment (girish946) [reply supertokens_python/recipe/emailverification/interfaces.py]: done

File: supertokens_python/recipe/emailverification/interfaces.py (hunk 21 above)
Comment (girish946) [reply supertokens_python/recipe/emailverification/interfaces.py:51]: done.

File: supertokens_python/recipe/emailverification/interfaces.py (hunk 19 above)
Comment (girish946) [reply supertokens_python/recipe/emailverification/interfaces.py]: done.

File: supertokens_python/recipe/emailverification/api/implementation.py (hunk 9 above)
Comment (girish946) [reply supertokens_python/recipe/emailverification/api/implementation.py]: fixed it.

File: supertokens_python/recipe/emailverification/api/implementation.py (hunk 8 above)
Comment (girish946) [reply supertokens_python/recipe/emailverification/api/implementation.py]: fixed.

File: supertokens_python/recipe/emailverification/api/implementation.py (hunk 6 above)
Comment (girish946) [reply supertokens_python/recipe/emailverification/api/implementation.py]: fixed.

**Comment by rishabhpoddar** [comment]
Closing this PR since it is combining many different tasks into one. Please create a new PR focused on the task in hand, in the supertokens repo itself.
